### PR TITLE
Cythonize `LatticePolytope.normal_form(algorithm='palp_native')`, change to default, add as a `Polyhedron` method

### PR DESCRIPTION
**Pull request by mkoeppe**
The current default algorithm, using the external, unmaintained `palp` package, is unreliable.

Here, as a follow-up on #35997, we make the reimplementation of the PALP algorithm in Python (`palp_native`) suitable as the new default by speeding it up.  

We also fix a bug in introduced in #35997.

We also make it available as a method of `Polyhedron` (for base ring `ZZ`).

@xuluze 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/groups/perm_gps/permgroup_element.pyx [hunk 1]
```diff
@@ -1298,6 +1298,39 @@ cdef class PermutationGroupElement(MultiplicativeGroupElement):
...
+            sage: s._transpose_left("b", "c")
+            ('a','e')('b','d','c')
+            sage: S(("b", "c")) * s
+            ('a','e')('b','d','c')
+        """
+        if j == k:
+            return self
+        cdef PermutationGroupElement prod = self._new_c()
+        cdef int i
+        for i from 0 <= i < self.n:
```
Comment (fchapoton) [inline src/sage/groups/perm_gps/permgroup_element.pyx]: could use a standard python-style loop

File: src/sage/geometry/polyhedron/base_ZZ.py [hunk 2]
```diff
@@ -834,3 +841,77 @@ def is_known_summand(poly):
...
+            ...
+            ValueError: normal form is not defined for unbounded polyhedra, got
+            A 2-dimensional polyhedron in ZZ^2 defined as the convex hull of 1 vertex and 2 rays
+
+        See :issue:`15280` for proposed extensions to these cases.
+        """
+        from sage.geometry.palp_normal_form import _palp_PM_max, _palp_canonical_order
+
+        if algorithm != "palp_native":
+            raise ValueError("algorithm must be 'palp_native'")
```
Comment (fchapoton) [inline src/sage/geometry/polyhedron/base_ZZ.py:914]: add a doctest for this path

File: src/sage/geometry/polyhedron/base_ZZ.py [hunk 3]
```diff
@@ -834,3 +841,77 @@ def is_known_summand(poly):
...
+
+        if not self.is_compact():
+            raise ValueError("normal form is not defined for unbounded polyhedra, got %s" % self)
+
+        PM = self.slack_matrix().transpose()
+        PM_max, permutations = _palp_PM_max(PM, check=True)
+        out = _palp_canonical_order(self.vertices(), PM_max, permutations)
+
+        if permutation:
+            return out
```
Comment (fchapoton) [inline src/sage/geometry/polyhedron/base_ZZ.py:927]: and for this path too

**Comment by mkoeppe** [comment]
All taken care of; any other comments?

File: src/sage/geometry/lattice_polytope.py [hunk 4]
```diff
@@ -3025,8 +3196,7 @@ def normal_form(self, algorithm="palp", permutation=False):
         elif algorithm == "palp_modified":
             result = self._palp_modified_normal_form(permutation=permutation)
         else:
-            raise ValueError('Algorithm must be palp, ' +
-                             'palp_native, or palp_modified.')
+            raise ValueError("algorithm must be 'palp', 'palp_native', or 'palp_modified'")
```
Comment (dimpase) [inline src/sage/geometry/lattice_polytope.py:3206]: Can you add a doctest for this ValueError?

**Review by dimpase — CHANGES_REQUESTED** [review]
apart from that missing trivial test, no problems

File: src/sage/geometry/lattice_polytope.py (hunk 4 above)
Comment (mkoeppe) [reply src/sage/geometry/lattice_polytope.py:3206]: Done in 0bd2a93f84e

**Review by dimpase — APPROVED** [review]
LGTM, thanks

**Comment by mkoeppe** [comment]
Thanks

**Comment by mkoeppe** [comment]
Marking this as a blocker in the hope it can sneak into 10.2 with the other open blocker

**Comment by tornaria** [comment]
> Marking this as a blocker in the hope it can sneak into 10.2 with the other open blocker

Why is this more worthy of "sneaking in" than e.g. #36551 ?

**Comment by vbraun** [comment]
Fails on 32-bit
```
**********************************************************************
File "src/sage/geometry/lattice_polytope.py", line 3141, in sage.geometry.lattice_polytope.LatticePolytopeClass.normal_form
Failed example:
    P.normal_form(algorithm="palp_modified")  # long time (22s)           # needs sage.groups
Exception raised:
    Traceback (most recent call last):
      File "sage/misc/cachefunc.pyx", line 1962, in sage.misc.cachefunc.CachedMethodCaller.__call__
        return cache[k]
    KeyError: (('palp_modified', False), ())
    During handling of the above exception, another exception occurred:
    Traceback (most recent call last):
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.geometry.lattice_polytope.LatticePolytopeClass.normal_form[14]>", line 1, in <module>
        P.normal_form(algorithm="palp_modified")  # long time (22s)           # needs sage.groups
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/misc/cachefunc.pyx", line 1967, in sage.misc.cachefunc.CachedMethodCaller.__call__
        w = self._instance_call(*args, **kwds)
      File "sage/misc/cachefunc.pyx", line 1842, in sage.misc.cachefunc.CachedMethodCaller._instance_call
        return self.f(self._instance, *args, **kwds)
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/geometry/lattice_polytope.py", line 3204, in normal_form
        result = self._palp_modified_normal_form(permutation=permutation)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/geometry/lattice_polytope.py", line 3259, in _palp_modified_normal_form
        PM_max = PM.permutation_normal_form()
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/matrix/matrix2.pyx", line 8637, in genexpr
        if not any(aN.is_permutation_of(j) for j in aM):
      File "sage/matrix/matrix2.pyx", line 8637, in genexpr
        if not any(aN.is_permutation_of(j) for j in aM):
      File "sage/matrix/matrix2.pyx", line 8792, in sage.matrix.matrix2.Matrix.is_permutation_of
        return N_B.is_isomorphic(M_B, certificate=False, edge_labels=True)
      File "/var/lib/buildbot/slave/sage_git/build/src/sage/graphs/generic_graph.py", line 23852, in is_isomorphic
        isom = isomorphic(GC, GC2, partition, partition2, (self._directed or self.has_loops()), 1)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/groups/perm_gps/partn_ref/refinement_graphs.pyx", line 166, in sage.groups.perm_gps.partn_ref.refinement_graphs.isomorphic
        cdef bint isomorphic = double_coset(<void *>GS1, <void *>GS2, part, ordering, n, &all_children_are_equivalent, &refine_by_degree, &compare_graphs, NULL, NULL, output)
      File "sage/groups/perm_gps/partn_ref/double_coset.pyx", line 367, in sage.groups.perm_gps.partn_ref.double_coset.double_coset
        raise MemoryError
    MemoryError
**********************************************************************
1 item had failures:
   1 of  17 in sage.geometry.lattice_polytope.LatticePolytopeClass.normal_form
    [667 tests, 1 failure, 16.73 s]
**********************************************************************
```

**Comment by mkoeppe** [comment]
Sorry about this; I've disabled this example now (it's also too long on 64 bit even for `# long`)

**Pull request merged**